### PR TITLE
Implementation of jailed shells

### PR DESCRIPTION
**Pull request by rjd22**
This is very work in progress and untested. The reason I am sending this PR is to gather some feedback before continuing. It's inspired by the following implementation: https://github.com/Skamasle/vestacp-jailkit-chroot/tree/master

I really miss the option to have a jailed shell to work in for things like running composer. This PR is meant to get the ball rolling and have a initial stab at offering a jailed bash shell.

When building this I found some issues with setting up the chroot on my Debian 12 installation hopefully you will be able to give me some insight what might go wrong.

Questions:
- I've separated the creation of a ssh jail from the sftp jail. Is this the right way to go, or do you rather have them combined because of the chroot creation?
- I've used the `ssh-jailed` group to move the user to the chroot directory. In sftp we use users. Would you rather have users for this or is a group okay? Maybe there is a really good reason users are used for sftp?
- ~~On my fresh debian 12 installation the following code doesn't seem to be executed even when the user is `nologin`. The directory /srv/jail/$user is not created and the systemd file is not created. It's quite hard to debug but might you know what could cause this? https://github.com/hestiacp/hestiacp/blob/main/func/main.sh#L1736-L1768~~

Note: I can imagine you all are really busy. Since I'm quite new and don't have much haste feel free to take your time.

### CODE REVIEW

File: bin/v-add-user
Comment (jaapmarcus) [inline bin/v-add-user]: Jailed ssh user  should probably not enabled by default It should be at least an option to be enabled / disabled on demand

File: bin/v-add-user-ssh-jail
Comment (jaapmarcus) [inline bin/v-add-user-ssh-jail]: If jail kit doens't exists we should build packages for it and ship it

File: bin/v-add-user
Comment (jaapmarcus) [reply bin/v-add-user]: On line: v-add-user-ssh-jail 41 only bash users allow to use it ... So it should be fine

**Review by jaapmarcus — COMMENTED** [review]
Haven't tested it yet but I assume we need to build the jailkit package aswell?

**Comment by rjd22** [comment]
@jaapmarcus still need to test it myself but chroot didn't seem to work properly on my server for some reason so I couldn't test it on my test server.

Jailkit can be installed by `apt install jailkit` but I didn't check yet how up to date it was. The whole reason that I didn't include that in the scripts is because it's a nice on/off switch.

File: bin/v-add-user-ssh-jail
Comment (rjd22) [reply bin/v-add-user-ssh-jail]: I still need to check how up-to-date the jailkit package is. But it's in the package manager. The reason I didn't include a auto install is because it allows for not enabling the feature when jailkit is not there.

**Comment by gOOvER** [comment]
> @jaapmarcus still need to test it myself but chroot didn't seem to work properly on my server for some reason so I couldn't test it on my test server.
> 
> Jailkit can be installed by `apt install jailkit` but I didn't check yet how up to date it was. The whole reason that I didn't include that in the scripts is because it's a nice on/off switch.

Debian 12 repo: 2.21
Debian Backports: 2.23

Ubuntu 22.04: 2.23

Lastest Version is: 2.23 on their homepage ([https://olivier.sessink.nl/jailkit/](https://olivier.sessink.nl/jailkit/))

**Comment by jaapmarcus** [comment]
> @jaapmarcus still need to test it myself but chroot didn't seem to work properly on my server for some reason so I couldn't test it on my test server.
> 
> Jailkit can be installed by `apt install jailkit` but I didn't check yet how up to date it was. The whole reason that I didn't include that in the scripts is because it's a nice on/off switch.

In that case we need to add it to depends line in: https://github.com/hestiacp/hestiacp/blob/bc41a3fd8a5845cd5309d873cb061ce14b7ccc17/src/deb/hestia/control#L9

Other wise existing installs will break... 

Also add it to the installer files...

File: bin/v-add-user-ssh-jail
Comment (jaapmarcus) [reply bin/v-add-user-ssh-jail]: But existing installs will break see:

https://drone.hestiacp.com/hestiacp/hestiacp/3535/1/5

We can't have it ..

Comment (rjd22) [reply bin/v-add-user-ssh-jail]: Ah ok, it doesn't like loud errors. I will remove the error messages. Jailkit not being installed should be an okay state so users have the option for non-jailed bash.

If you have a suggestion on how to support jailed with non-jailed bash in one installation I'm all ears. I couldn't think of one ^^°.

**Comment by rjd22** [comment]
> > @jaapmarcus still need to test it myself but chroot didn't seem to work properly on my server for some reason so I couldn't test it on my test server.
> > Jailkit can be installed by `apt install jailkit` but I didn't check yet how up to date it was. The whole reason that I didn't include that in the scripts is because it's a nice on/off switch.
> 
> In that case we need to add it to depends line in:
> 
> https://github.com/hestiacp/hestiacp/blob/bc41a3fd8a5845cd5309d873cb061ce14b7ccc17/src/deb/hestia/control#L9
> 
> Other wise existing installs will break...
> 
> Also add it to the installer files...

@jaapmarcus I've added some starting tests to check if the bash is properly jailed. I will add some more assertions for this but somehow I'm unable to get the install of the jailkit package to work correctly. I'm sure I'm missing something but cannot figure out where I'm missing the dependency:

https://github.com/hestiacp/hestiacp/pull/4052/commits/464513fbeb4b7dcb8864641c0c53c0391bed1ee1

**Comment by rjd22** [comment]
Update: I've started testing this on a development environment and still have some issues I need to figure out.

Because bash is the default shell the users admin and ubuntu are jailed after install, breaking the installation.

Avoiding ubuntu to be jailed is easy (just check if directory /home/user/conf/web exist). But avoiding admin to be jailed is harder.

If someone has ideas on how to reliably figure out if the user is a non admin user I would love to hear so.

File: bin/v-add-user
Comment (rjd22) [reply bin/v-add-user]: Note to self: Flag check still needs to be added here.

**Comment by rjd22** [comment]
A lot of work on this has been done. I've added the ability to enable a jailed shell for a package and user. Also the jail now works with multiple shells. 

I only tested bash shell for now so some extra guards for stupid user configurations might be in order (combining nologin or rssh with a jail).

It would be great to hear what you think. I could also use a pointer on how to get jailkit installed on the build server for testing.

**Comment by jaapmarcus** [comment]
> I could also use a pointer on how to get jailkit installed on the build server for testing.

It needs to be done by me or somebody with access to the build servers via ssh... As we install the packages via dpkg  installing additional packages are not possible ...

**Comment by rjd22** [comment]
> > I could also use a pointer on how to get jailkit installed on the build server for testing.
> 
> It needs to be done by me or somebody with access to the build servers via ssh... As we install the packages via dpkg installing additional packages are not possible ...

Ah ok, what would be the correct way to go about this? The reason the tests fail is because I added jailkit to the `depends` config value in `deb/hestia/control`.

**Comment by jaapmarcus** [comment]
> > > I could also use a pointer on how to get jailkit installed on the build server for testing.
> > 
> > 
> > It needs to be done by me or somebody with access to the build servers via ssh... As we install the packages via dpkg installing additional packages are not possible ...
> 
> Ah ok, what would be the correct way to go about this? The reason the tests fail is because I added jailkit to the `depends` config value in `deb/hestia/control`.

Wait till I have time to fix it .. :)

Just ran apt --fix-broken install

**Comment by rjd22** [comment]
I've finished the implementation and an waiting the tests to finish. Here is an overview of the functionality that is included:

## Features
* Ability to set if the shell needs to be jailed in packages
* Ability to set if the shell needs to be jailed for users
* Creating a user from a package with the jailed option enabled will result in a jailed user

## Jail functionality
* Internet connectivity
* SSH
* SCP
* SFTP
* Git
* Perl
* PHP
* Composer (if composer installed with `v-add-user-composer`)
* Net utils: wget, lynx, ftp, host, rsync, smbclient
* Basic Shell: /bin/sh, bash, ls, cat, chmod, mkdir, cp, cpio, date, dd, echo, egrep, false, fgrep, grep, gunzip, gzip, ln, ls, mkdir, mktemp, more, mv, pwd, rm, rmdir, sed, sh, sleep, sync, tar, touch, true, uncompress, zcat
* Midnight Commander: mc, mcedit, mcview, /usr/share/mc
* Extended Shell: awk, bzip2, bunzip2, ldd, less, clear, cut, du, find, head, less, md5sum, nice, sort, tac, tail, tr, sort, wc, watch, whoami
* Editors: vim, joe and nano

## Breaking changes
* I had to move the jailed homedir to make bash and composer work correctly. The directory was moved from `/srv/jail/$user/home` to `/srv/jail/$user/home/$user`

## Testing done by me
* Adding and editing packages with jail enabled / disabled
* Adding and editing users with jail enabled / disabled
* SFTP only with nologin works correctly
* SSH bash jailed works correctly
* SSH bash normal works correctly
* Within the jail the ability to use PHP and Composer

I'm really interested in what you think. This was more work than I initially thought >_<.

**Comment by jaapmarcus** [comment]
> * PHP

We need include all versions 5.6 -> 8.2 if installed 
> * Composer (if composer installed with `v-add-user-composer`)

wp cli...

**Comment by rjd22** [comment]
> > * PHP
> 
> We need include all versions 5.6 -> 8.2 if installed

I've added config files to install all php versions if available on the server. I've used php8.2 as default PHP because afaik that seems to be the default version (not that it matters since all versions are installed now).

> > * Composer (if composer installed with `v-add-user-composer`)
> 
> wp cli...

WP Cli works correctly.

**Comment by rjd22** [comment]
The builds now passes. Hopefully someone feels up to the task to review this. (⁠๑⁠•⁠﹏⁠•⁠)

**Comment by rjd22** [comment]
Most likely you're all very busy. Would someone be able to review this and if this is complete or still needs some changes?

The reason I'm asking is that if we can get this in before the 1.9 release I can remove some of the chroot lines that repair the home directory location changes because AFAIK those changes were not released yet. 

I kept them in because I want you all to be able to automatically fix your dev boxes before removing the lines.

How do you all feel about the group name btw? I didn't think deeply about it but maybe you rather have it prefixed with hestia like `hestia-ssh-jailed` or `hestia-jailed-ssh` (with or without dashes). 

Would you also maybe like the sftp to use groups so we don't have the change `sshd_config` for every users? Like `hestia-sftp-only`.

**Comment by jaapmarcus** [comment]
> Most likely you're all very busy. Would someone be able to review this and if this is complete or still needs some changes?
> 
> The reason I'm asking is that if we can get this in before the 1.9 release I can remove some of the chroot lines that repair the home directory location changes because AFAIK those changes were not released yet.
> 
> I kept them in because I want you all to be able to automatically fix your dev boxes before removing the lines.
> 
> How do you all feel about the group name btw? I didn't think deeply about it but maybe you rather have it prefixed with hestia like `hestia-ssh-jailed` or `hestia-jailed-ssh` (with or without dashes).
> 
> Would you also maybe like the sftp to use groups so we don't have the change `sshd_config` for every users? Like `hestia-sftp-only`.

We had  major security leak in Hestia that needed to be patched ...  So was quite busy with it ...  Sorry.. Will look it later today...

**Comment by rjd22** [comment]
I'm going to give a small ping again. Hopefully someone can take a look at this?

File: web/templates/pages/edit_user.php
Comment (jaapmarcus) [inline web/templates/pages/edit_user.php:224]: I don't know if we should run v-add-ssh-jail on default on existing systems but we don't do it we probally should not display the option

As it is installed anyway on upgrade it probally doesn't hurt to do enable it on update by default

File: install/common/jailkit/jk_init.ini
Comment (jaapmarcus) [inline install/common/jailkit/jk_init.ini]: unzip 
zip 
and zstd should also be added. 


For /usr/lib/locale/en_US.utf8 how are we going to handle non en_US locales?

File: install/common/jailkit/jk_init.ini
Comment (jaapmarcus) [inline install/common/jailkit/jk_init.ini:173]: Fan to include npm to the list used in many software development

File: bin/v-add-user-ssh-jail
Comment (jaapmarcus) [inline bin/v-add-user-ssh-jail:62]: npm should be added

File: bin/v-add-user-ssh-jail
Comment (jaapmarcus) [inline bin/v-add-user-ssh-jail:64]: After first time running this command I did get:

```
usermod: user eris is currently used by process 75716
failed to execute usermod -d /srv/jail/eris/./home/eris -s /usr/sbin/jk_chrootsh eris
failed to modify user eris
usermod: no changes
usermod: no changes
```
This might be caused that I was already logged in into via ssh

How ever everything I try to login I get the error: 

Last login: Sun Nov  5 21:53:04 2023 from xxxxx
-bash: groups: command not found

File: bin/v-change-user-shell
Comment (jaapmarcus) [inline bin/v-change-user-shell]: Default value should be no probally unless otherwise specified..

File: bin/v-list-user
Comment (jaapmarcus) [inline bin/v-list-user]: Please keep the spacing intact :)

File: docs/docs/reference/cli.md
Comment (jaapmarcus) [inline docs/docs/reference/cli.md]: Please remove the changes will update the CLI docs after it has been merged

**Review by jaapmarcus — CHANGES_REQUESTED** [review]
Sorry for the huge delay

see remarks it seems to work fine :)

File: install/common/jailkit/jk_init.ini
Comment (rjd22) [reply install/common/jailkit/jk_init.ini:173]: The reasoning I had not to added it is that most likely people will use NPM to compile their assets wasting a lot of system resources (it maxes out the CPU). 

It felt unsafe to make it a default capability and people should compile their assets outside of the shared hosting IMO.

If you feel differently I'll gladly add it.

File: bin/v-add-user-ssh-jail
Comment (rjd22) [reply bin/v-add-user-ssh-jail:64]: This indeed has to do that you're already logged in. There is nothing to be done in that situation but honestly I don't think it will happen often and can be easily fixed by changing the shell to unjailed and then to jailed again.

That chances of it happening is quite low. Normally the setting is set correctly on user creation or it's changed from nologin to a shell with jail.

Only when you change a unjailed shell to a jailed shell this will happen.

File: install/common/jailkit/jk_init.ini
Comment (rjd22) [reply install/common/jailkit/jk_init.ini]: This is the default config of jailkit. I'm not sure if we can just copy in other locales and have them work so that would need more experimentation.

Honestly I would wait to see if there is a specific need for other locales after some people start using it.

File: web/templates/pages/edit_user.php
Comment (rjd22) [reply web/templates/pages/edit_user.php:224]: What would you prefer? From security perspective I would say that installing jailkit and offering the option after updating would be preferred so people can start jailing their users and lower the amount of access their users have trough ssh.

But as a risk assessment I can understand that you rather have this a feature that needs to be enabled before use.

File: bin/v-add-user-ssh-jail
Comment (jaapmarcus) [reply bin/v-add-user-ssh-jail:64]: Still this error happens even after enabling / disabling it ...

Last login: Sun Nov 5 21:53:04 2023 from xxxxx
-bash: groups: command not found

Comment (rjd22) [reply bin/v-add-user-ssh-jail:64]: This is because we're unable to change the passwd and groups file within the jail and thus cannot assign the right homedir and shell within the jail if there is an active ssh session.

The only solution I can think of is disconnecting the user if we change the jail if there is a current session.

Comment (jaapmarcus) [reply bin/v-add-user-ssh-jail:64]: > This is because we're unable to change the passwd and groups file within the jail and thus cannot assign the right homedir and shell within the jail if there is an active ssh session.
> 
> The only solution I can think of is disconnecting the user if we change the jail if there is a current session.

This was without an active ssh session (Except for the root user as well)

Comment (rjd22) [reply bin/v-add-user-ssh-jail:64]: Ok, I will check it out.

File: install/common/jailkit/jk_init.ini
Comment (rjd22) [reply install/common/jailkit/jk_init.ini:173]: @jaapmarcus what do you think of the above? Do you want to add npm anyway?

File: bin/v-add-user-ssh-jail
Comment (rjd22) [reply bin/v-add-user-ssh-jail:64]: This should be fixed now. I added groups and made sure all groups are set before jailing the user

File: install/common/jailkit/jk_init.ini
Comment (rjd22) [reply install/common/jailkit/jk_init.ini:173]: Added the NPM config but didn't enable it yet by default because I want to discuss the performance implications above before enabling it by default.

File: install/common/jailkit/jk_init.ini
Comment (rjd22) [reply install/common/jailkit/jk_init.ini]: Added the packages

**Comment by rjd22** [comment]
@jaapmarcus would you be able to re-review and check the above questions?

**Comment by rjd22** [comment]
@jaapmarcus could you take another look at the comments and fixes? 

I'm planning some more improvements but want to get this big one out of the way first.

**Comment by rjd22** [comment]
I understand you're all really busy but can someone please take a look at this PR so that I can continue the work? There are a few questions open and if there is no objection we should be able to get this working.

**Comment by rjd22** [comment]
Would someone be willing to review this please? A lot of work has gone into it and it would be a shame if we couldn't include it.

The NPM config has been added but not enabled by default yet for the reason of being able to easily eat up the whole server CPU by building applications with it. I would say it might be wise to see if we could limit those resources before enabling it by default.

**Comment by rjd22** [comment]
@jaapmarcus @AlecRust Is someone able to atleast confirm the above? I've been waiting for a response for 2 months now. I know your not obliged to respond to PR's but in that case I'll take the hint and move on from this project. :)

**Comment by ScIT-Raphael** [comment]
Hi @rjd22 

Sorry for the long delay, we currently have not much persons that are able to invest their time into hestia. I'll have a look at the communication and will try to provide a code review until end of this week.

Maybe you want to join our discord, so you would have a more direct way of communication - please let me know when you've joined and provide also your username, so I can adjust your permissions: https://discord.com/invite/4VYq26s5

**Comment by rjd22** [comment]
@ScIT-Raphael thank you for your message. I joined the discord and should have the same username as here "rjd22".

**Review by jaapmarcus — APPROVED** [review]

**Comment by jaapmarcus** [comment]
We should enable it on new installs on default...

**Review by ScIT-Raphael — APPROVED** [review]

**Comment by jaapmarcus** [comment]
Thanks :heart:

**Pull request merged**
